src/rgw/rgw_sal_motr.cc

```diff
@@ -1481,12 +1472,18 @@ int MotrObject::MotrReadOp::read(int64_t off, int64_t end, bufferlist& bl, optio
 int MotrObject::MotrReadOp::iterate(const DoutPrefixProvider* dpp, int64_t off, int64_t end, RGWGetDataCB* cb, optional_yield y)
 {
   int rc;
-
-  if (source->category == RGWObjCategory::MultiMeta)
+  // Open and read the object here.
+  if (source->category == RGWObjCategory::MultiMeta){
+    ldpp_dout(dpp, 20) <<__func__<< ": open obj parts..." << dendl;
+    source->get_part_objs(dpp, this->part_objs)? : source->open_part_objs(dpp, this->part_objs);
+    ldpp_dout(dpp, 20) <<__func__<< ": read multipart obj parts..." << dendl;
     rc = source->read_multipart_obj(dpp, off, end, cb, part_objs);
-  else
+  } else {
+    ldpp_dout(dpp, 20) <<__func__<< ": open object..." << dendl;
+    source->open_mobj(dpp);
```

Comment:
Why should we open it on each `iterate()` call?

Reply:
@pranavPawar07  Does iterate()  gets called multiple times? as Andry mentioned in the PR comment
We need to test get-object api as well with small and large objects and check if iterate is called multiple times.

Reply:
no, iterate() is called only once. I have tested the get-object API as well, for both small and large size objects, working as expected.

Reply:
@pranavPawar07, how did you check this? Please, provide some evidence here.

Reply:
Below logs are for get-object s3api request on 1GB object.
If you check closely, iterate() is called only once.

```
0600 7fe54befc700 20 req 0 0.096000433s s3:get_obj prepare: object's etag: e5c834fbdaa6bfd8eac5eb9404eefdd4
2022-04-04T05:11:38.716-0600 7fe54befc700 15 req 0 0.096000433s Encryption mode:
2022-04-04T05:11:38.716-0600 7fe54befc700 20 req 0 0.096000433s s3:get_obj iterate: open object...
2022-04-04T05:11:38.716-0600 7fe54befc700 20 req 0 0.096000433s s3:get_obj open_mobj: oid=4401001:e37a000002900000
2022-04-04T05:11:38.716-0600 7fe54befc700 20 req 0 0.096000433s s3:get_obj open_mobj: rc=0
2022-04-04T05:11:38.716-0600 7fe54befc700 20 req 0 0.096000433s s3:get_obj MotrObject::read_mobj(): off=0 end=1048576000
2022-04-04T05:11:38.716-0600 7fe54befc700 20 req 0 0.096000433s s3:get_obj MotrObject::read_mobj(): bs=41943040
2022-04-04T05:11:38.716-0600 7fe54befc700 20 req 0 0.096000433s s3:get_obj MotrObject::read_mobj(): off=0 actual=41943040
2022-04-04T05:11:38.716-0600 7fe54befc700 20 req 0 0.096000433s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:39.048-0600 7fe54befc700 20 req 0 0.428001910s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:39.151-0600 7fe53c6dd700 20 req 0 0.531002402s s3:get_obj MotrObject::read_mobj(): off=41943040 actual=41943040
2022-04-04T05:11:39.151-0600 7fe53c6dd700 20 req 0 0.531002402s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:39.293-0600 7fe53c6dd700 20 req 0 0.673003018s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:39.402-0600 7fe530ec6700 20 req 0 0.782003522s s3:get_obj MotrObject::read_mobj(): off=83886080 actual=41943040
2022-04-04T05:11:39.402-0600 7fe530ec6700 20 req 0 0.782003522s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:39.537-0600 7fe530ec6700 20 req 0 0.917004108s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:39.643-0600 7fe522eaa700 20 req 0 1.023004651s s3:get_obj MotrObject::read_mobj(): off=125829120 actual=41943040
2022-04-04T05:11:39.643-0600 7fe522eaa700 20 req 0 1.023004651s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:39.765-0600 7fe522eaa700 20 req 0 1.145005226s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:39.891-0600 7fe519e98700 20 req 0 1.271005750s s3:get_obj MotrObject::read_mobj(): off=167772160 actual=41943040
2022-04-04T05:11:39.891-0600 7fe519e98700 20 req 0 1.271005750s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:40.037-0600 7fe519e98700 20 req 0 1.417006373s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:40.146-0600 7fe50f683700 20 req 0 1.526006937s s3:get_obj MotrObject::read_mobj(): off=209715200 actual=41943040
2022-04-04T05:11:40.146-0600 7fe50f683700 20 req 0 1.526006937s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:40.253-0600 7fe50f683700 20 req 0 1.633007407s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:40.339-0600 7fe504e6e700 20 req 0 1.719007730s s3:get_obj MotrObject::read_mobj(): off=251658240 actual=41943040
2022-04-04T05:11:40.340-0600 7fe504e6e700 20 req 0 1.720007658s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:40.448-0600 7fe504e6e700 20 req 0 1.828008175s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:40.541-0600 7fe4f8655700 20 req 0 1.921008587s s3:get_obj MotrObject::read_mobj(): off=293601280 actual=41943040
2022-04-04T05:11:40.542-0600 7fe4f8655700 20 req 0 1.922008753s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:40.731-0600 7fe4f8655700 20 req 0 2.111009598s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:40.832-0600 7fe4ede40700 20 req 0 2.212009907s s3:get_obj MotrObject::read_mobj(): off=335544320 actual=41943040
2022-04-04T05:11:40.832-0600 7fe4ede40700 20 req 0 2.212009907s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:40.947-0600 7fe4ede40700 20 req 0 2.327010632s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:41.045-0600 7fe4e3e2c700 20 req 0 2.425010920s s3:get_obj MotrObject::read_mobj(): off=377487360 actual=41943040
2022-04-04T05:11:41.045-0600 7fe4e3e2c700 20 req 0 2.425010920s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:41.153-0600 7fe4e3e2c700 20 req 0 2.533011436s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:41.250-0600 7fe4dae1a700 20 req 0 2.630011797s s3:get_obj MotrObject::read_mobj(): off=419430400 actual=41943040
2022-04-04T05:11:41.250-0600 7fe4dae1a700 20 req 0 2.630011797s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:41.399-0600 7fe4dae1a700 20 req 0 2.779012680s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:41.523-0600 7fe4d0605700 20 req 0 2.903013229s s3:get_obj MotrObject::read_mobj(): off=461373440 actual=41943040
2022-04-04T05:11:41.523-0600 7fe4d0605700 20 req 0 2.903013229s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:41.673-0600 7fe4d0605700 20 req 0 3.053013802s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:41.775-0600 7fe4c35eb700 20 req 0 3.155014038s s3:get_obj MotrObject::read_mobj(): off=503316480 actual=41943040
2022-04-04T05:11:41.775-0600 7fe4c35eb700 20 req 0 3.155014038s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:41.929-0600 7fe4c35eb700 20 req 0 3.309014797s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:42.020-0600 7fe4b8dd6700 20 req 0 3.400015354s s3:get_obj MotrObject::read_mobj(): off=545259520 actual=41943040
2022-04-04T05:11:42.021-0600 7fe4b8dd6700 20 req 0 3.401015282s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:42.190-0600 7fe4b8dd6700 20 req 0 3.570015907s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:42.288-0600 7fe4addc0700 20 req 0 3.668016434s s3:get_obj MotrObject::read_mobj(): off=587202560 actual=41943040
2022-04-04T05:11:42.288-0600 7fe4addc0700 20 req 0 3.668016434s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:42.422-0600 7fe4addc0700 20 req 0 3.802016973s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:42.517-0600 7fe5a1fa8700 20 req 0 3.897017717s s3:get_obj MotrObject::read_mobj(): off=629145600 actual=41943040
2022-04-04T05:11:42.517-0600 7fe5a1fa8700 20 req 0 3.897017717s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:42.614-0600 7fe5a1fa8700 20 req 0 3.994018078s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:42.724-0600 7fe596f92700 20 req 0 4.104018211s s3:get_obj MotrObject::read_mobj(): off=671088640 actual=41943040
2022-04-04T05:11:42.725-0600 7fe596f92700 20 req 0 4.105018139s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:42.853-0600 7fe596f92700 20 req 0 4.233019352s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:42.953-0600 7fe58e781700 20 req 0 4.333019733s s3:get_obj MotrObject::read_mobj(): off=713031680 actual=41943040
2022-04-04T05:11:42.954-0600 7fe58e781700 20 req 0 4.334019661s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:43.139-0600 7fe58e781700 20 req 0 4.519020557s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:43.236-0600 7fe581767700 20 req 0 4.616021156s s3:get_obj MotrObject::read_mobj(): off=754974720 actual=41943040
2022-04-04T05:11:43.236-0600 7fe581767700 20 req 0 4.616021156s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:43.363-0600 7fe581767700 20 req 0 4.743021488s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:43.438-0600 7fe576f52700 20 req 0 4.818021774s s3:get_obj MotrObject::read_mobj(): off=796917760 actual=41943040
2022-04-04T05:11:43.439-0600 7fe576f52700 20 req 0 4.819021702s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:43.534-0600 7fe576f52700 20 req 0 4.914021969s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:43.622-0600 7fe56e741700 20 req 0 5.002022266s s3:get_obj MotrObject::read_mobj(): off=838860800 actual=41943040
2022-04-04T05:11:43.622-0600 7fe56e741700 20 req 0 5.002022266s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:43.763-0600 7fe56e741700 20 req 0 5.143023014s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:43.851-0600 7fe560f26700 20 req 0 5.231023788s s3:get_obj MotrObject::read_mobj(): off=880803840 actual=41943040
2022-04-04T05:11:43.851-0600 7fe560f26700 20 req 0 5.231023788s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:44.175-0600 7fe560f26700 20 req 0 5.555025101s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:44.266-0600 7fe556711700 20 req 0 5.647025108s s3:get_obj MotrObject::read_mobj(): off=922746880 actual=41943040
2022-04-04T05:11:44.267-0600 7fe556711700 20 req 0 5.647025108s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:44.394-0600 7fe556711700 20 req 0 5.774025917s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:44.475-0600 7fe54aefa700 20 req 0 5.855026245s s3:get_obj MotrObject::read_mobj(): off=964689920 actual=41943040
2022-04-04T05:11:44.475-0600 7fe54aefa700 20 req 0 5.855026245s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:44.580-0600 7fe54aefa700 20 req 0 5.960026741s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:44.676-0600 7fe540ee6700 20 req 0 6.056026936s s3:get_obj MotrObject::read_mobj(): off=1006632960 actual=41943040
2022-04-04T05:11:44.676-0600 7fe540ee6700 20 req 0 6.056026936s s3:get_obj MotrObject::read_mobj(): init read op rc=0
2022-04-04T05:11:44.820-0600 7fe540ee6700 20 req 0 6.200027466s s3:get_obj MotrObject::read_mobj(): call cb to process data
2022-04-04T05:11:44.911-0600 7fe5366d1700 2 req 0 6.291028500s s3:get_obj completing
2022-04-04T05:11:44.911-0600 7fe5366d1700 2 req 0 6.291028500s s3:get_obj op status=0
2022-04-04T05:11:44.911-0600 7fe5366d1700 2 req 0 6.291028500s s3:get_obj http status=200
2022-04-04T05:11:44.911-0600 7fe5366d1700 1 ====== req done req=0x7fe5c2eba780 op status=0 http_status=200 latency=6.291028500s ======
2022-04-04T05:11:44.912-0600 7fe5366d1700 1 beast: 0x7fe5c2eba780: ::1 - mgwuser [04/Apr/2022:05:11:38.620 -0600] "GET /test/1gb HTTP/1.1" 200 1048576000 - "aws-cli/1.22.71 Python/3.6.8 Linux/4.18.0-305.3.1.el8_4.x86_64 botocore/1.24.16" - latency=6.291028500s
2022-04-04T05:11:44.932-0600 7fe535ed0700 20 failed to read header: end of stream
```

Reply:
I see.